File: pkg/ui/src/api/services/stepfunctions.ts
/**
 * Step Functions Service API Client
 * REST HTTP client for Step Functions via Go proxy
 * @module api/services/stepfunctions
 */

import { PROXY_BACKEND } from '@/config'
import { APIError } from '../client'

const api = PROXY_BACKEND.replace(/\/$/, '')

// --- State Machines ---

export async function listStateMachines(): Promise<any> {
  const res = await fetch(`${api}/step-functions/state-machines`)
  if (!res.ok) throw new APIError(`List state machines failed`, res.status, 'stepfunctions')
  return res.json()
}

export async function createStateMachine(body: Record<string, unknown>): Promise<any> {
  const res = await fetch(`${api}/step-functions/state-machines`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
  if (!res.ok) throw new APIError(`Create state machine failed`, res.status, 'stepfunctions')
  return res.json()
}

export async function describeStateMachine(arn: string): Promise<any> {
  const res = await fetch(`${api}/step-functions/state-machines/${encodeURIComponent(arn)}`)
  if (!res.ok) throw new APIError(`Describe state machine failed`, res.status, 'stepfunctions')
  return res.json()
}

export async function updateStateMachine(arn: string, body: Record<string, unknown>): Promise<any> {
  const res = await fetch(`${api}/step-functions/state-machines/${encodeURIComponent(arn)}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
  if (!res.ok) throw new APIError(`Update state machine failed`, res.status, 'stepfunctions')
  return res.json()
}

export async function deleteStateMachine(arn: string): Promise<void> {
  const res = await fetch(`${api}/step-functions/state-machines/${encodeURIComponent(arn)}`, { method: 'DELETE' })
  if (!res.ok) throw new APIError(`Delete state machine failed`, res.status, 'stepfunctions')
}

// --- Executions ---

export async function startExecution(arn: string, body: Record<string, unknown>): Promise<any> {
  const res = await fetch(`${api}/step-functions/state-machines/${encodeURIComponent(arn)}/executions`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
  if (!res.ok) throw new APIError(`Start execution failed`, res.status, 'stepfunctions')
  return res.json()
}

export async function listExecutions(arn: string, params?: Record<string, unknown>): Promise<any> {
  const query = params ? '?' + new URLSearchParams(
    Object.entries(params).filter(([_, v]) => v !== undefined).map(([k, v]) => [k, String(v)])
  ).toString() : ''
  const res = await fetch(`${api}/step-functions/state-machines/${encodeURIComponent(arn)}/executions${query}`)
  if (!res.ok) throw new APIError(`List executions failed`, res.status, 'stepfunctions')
  return res.json()
}

export async function stopExecution(arn: string, executionArn: string, body: Record<string, unknown>): Promise<void> {
  const res = await fetch(`${api}/step-functions/executions/${encodeURIComponent(executionArn)}/stop`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
  if (!res.ok) throw new APIError(`Stop execution failed`, res.status, 'stepfunctions')
}

export async function describeExecution(arn: string, executionArn: string): Promise<any> {
  const res = await fetch(`${api}/step-functions/executions/${encodeURIComponent(executionArn)}`)
  if (!res.ok) throw new APIError(`Describe execution failed`, res.status, 'stepfunctions')
  return res.json()
}

export async function getExecutionHistory(arn: string, executionArn: string, params?: Record<string, unknown>): Promise<any> {
  const query = params ? '?' + new URLSearchParams(
    Object.entries(params).filter(([_, v]) => v !== undefined).map(([k, v]) => [k, String(v)])
  ).toString() : ''
  const res = await fetch(`${api}/step-functions/executions/${encodeURIComponent(executionArn)}/history${query}`)
  if (!res.ok) throw new APIError(`Get execution history failed`, res.status, 'stepfunctions')
  return res.json()
}

File: pkg/proxy/internal/adapters/http/stepfunctions.go
package httphandlers

import (
	"context"
	"net/http"

	"github.com/aws/aws-sdk-go-v2/service/sfn"
	"github.com/gin-gonic/gin"
)

func (h *ProxyHandler) HandleListStateMachines(c *gin.Context) {
	bodyBytes := readBody(c)
	ctx := context.Background()
	input := &sfn.ListStateMachinesInput{}
	if err := parseBody(c, bodyBytes, input); err != nil {
		sendError(c, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.svc.StepFunctions().ListStateMachines(ctx, input)
	if err != nil {
		sendError(c, http.StatusInternalServerError, "Failed to list state machines", err)
		return
	}
	c.JSON(http.StatusOK, result)
}

func (h *ProxyHandler) HandleCreateStateMachine(c *gin.Context) {
	bodyBytes := readBody(c)
	ctx := context.Background()
	input := &sfn.CreateStateMachineInput{}
	if err := parseBody(c, bodyBytes, input); err != nil {
		sendError(c, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	result, err := h.svc.StepFunctions().CreateStateMachine(ctx, input)
	if err != nil {
		sendError(c, http.StatusInternalServerError, "Failed to create state machine", err)
		return
	}
	c.JSON(http.StatusCreated, result)
}

func (h *ProxyHandler) HandleDescribeStateMachine(c *gin.Context) {
	bodyBytes := readBody(c)
	ctx := context.Background()
	input := &sfn.DescribeStateMachineInput{}
	if err := parseBody(c, bodyBytes, input); err != nil {
		sendError(c, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	// Allow ARN from path param if not in body
	if input.StateMachineArn == nil || *input.StateMachineArn == "" {
		arn := c.Param("arn")
		if arn != "" {
			input.StateMachineArn = &arn
		}
	}
	result, err := h.svc.StepFunctions().DescribeStateMachine(ctx, input)
	if err != nil {
		sendError(c, http.StatusInternalServerError, "Failed to describe state machine", err)
		return
	}
	c.JSON(http.StatusOK, result)
}

func (h *ProxyHandler) HandleUpdateStateMachine(c *gin.Context) {
	bodyBytes := readBody(c)
	ctx := context.Background()
	input := &sfn.UpdateStateMachineInput{}
	if err := parseBody(c, bodyBytes, input); err != nil {
		sendError(c, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	if input.StateMachineArn == nil || *input.StateMachineArn == "" {
		arn := c.Param("arn")
		if arn != "" {
			input.StateMachineArn = &arn
		}
	}
	result, err := h.svc.StepFunctions().UpdateStateMachine(ctx, input)
	if err != nil {
		sendError(c, http.StatusInternalServerError, "Failed to update state machine", err)
		return
	}
	c.JSON(http.StatusOK, result)
}

func (h *ProxyHandler) HandleDeleteStateMachine(c *gin.Context) {
	bodyBytes := readBody(c)
	ctx := context.Background()
	input := &sfn.DeleteStateMachineInput{}
	if err := parseBody(c, bodyBytes, input); err != nil {
		sendError(c, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	if input.StateMachineArn == nil || *input.StateMachineArn == "" {
		arn := c.Param("arn")
		if arn != "" {
			input.StateMachineArn = &arn
		}
	}
	result, err := h.svc.StepFunctions().DeleteStateMachine(ctx, input)
	if err != nil {
		sendError(c, http.StatusInternalServerError, "Failed to delete state machine", err)
		return
	}
	c.JSON(http.StatusOK, result)
}

func (h *ProxyHandler) HandleStartExecution(c *gin.Context) {
	bodyBytes := readBody(c)
	ctx := context.Background()
	input := &sfn.StartExecutionInput{}
	if err := parseBody(c, bodyBytes, input); err != nil {
		sendError(c, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	if input.StateMachineArn == nil || *input.StateMachineArn == "" {
		arn := c.Param("arn")
		if arn != "" {
			input.StateMachineArn = &arn
		}
	}
	result, err := h.svc.StepFunctions().StartExecution(ctx, input)
	if err != nil {
		sendError(c, http.StatusInternalServerError, "Failed to start execution", err)
		return
	}
	c.JSON(http.StatusOK, result)
}

func (h *ProxyHandler) HandleListExecutions(c *gin.Context) {
	bodyBytes := readBody(c)
	ctx := context.Background()
	input := &sfn.ListExecutionsInput{}
	if err := parseBody(c, bodyBytes, input); err != nil {
		sendError(c, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	if input.StateMachineArn == nil || *input.StateMachineArn == "" {
		arn := c.Param("arn")
		if arn != "" {
			input.StateMachineArn = &arn
		}
	}
	result, err := h.svc.StepFunctions().ListExecutions(ctx, input)
	if err != nil {
		sendError(c, http.StatusInternalServerError, "Failed to list executions", err)
		return
	}
	c.JSON(http.StatusOK, result)
}

func (h *ProxyHandler) HandleStopExecution(c *gin.Context) {
	bodyBytes := readBody(c)
	ctx := context.Background()
	input := &sfn.StopExecutionInput{}
	if err := parseBody(c, bodyBytes, input); err != nil {
		sendError(c, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	if input.ExecutionArn == nil || *input.ExecutionArn == "" {
		executionArn := c.Param("executionArn")
		if executionArn != "" {
			input.ExecutionArn = &executionArn
		}
	}
	result, err := h.svc.StepFunctions().StopExecution(ctx, input)
	if err != nil {
		sendError(c, http.StatusInternalServerError, "Failed to stop execution", err)
		return
	}
	c.JSON(http.StatusOK, result)
}

func (h *ProxyHandler) HandleDescribeExecution(c *gin.Context) {
	bodyBytes := readBody(c)
	ctx := context.Background()
	input := &sfn.DescribeExecutionInput{}
	if err := parseBody(c, bodyBytes, input); err != nil {
		sendError(c, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	if input.ExecutionArn == nil || *input.ExecutionArn == "" {
		executionArn := c.Param("executionArn")
		if executionArn != "" {
			input.ExecutionArn = &executionArn
		}
	}
	result, err := h.svc.StepFunctions().DescribeExecution(ctx, input)
	if err != nil {
		sendError(c, http.StatusInternalServerError, "Failed to describe execution", err)
		return
	}
	c.JSON(http.StatusOK, result)
}

func (h *ProxyHandler) HandleGetExecutionHistory(c *gin.Context) {
	bodyBytes := readBody(c)
	ctx := context.Background()
	input := &sfn.GetExecutionHistoryInput{}
	if err := parseBody(c, bodyBytes, input); err != nil {
		sendError(c, http.StatusBadRequest, "Invalid request body", err)
		return
	}
	if input.ExecutionArn == nil || *input.ExecutionArn == "" {
		executionArn := c.Param("executionArn")
		if executionArn != "" {
			input.ExecutionArn = &executionArn
		}
	}
	result, err := h.svc.StepFunctions().GetExecutionHistory(ctx, input)
	if err != nil {
		sendError(c, http.StatusInternalServerError, "Failed to get execution history", err)
		return
	}
	c.JSON(http.StatusOK, result)
}
